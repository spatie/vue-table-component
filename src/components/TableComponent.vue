<template>
    <div class="table-component">
        <div v-if="showFilter && filterableColumnExists" class="table-component__filter">
            <input
                    :class="fullFilterInputClass"
                    type="text"
                    v-model="filter"
                    :placeholder="filterPlaceholder"
            >
            <a
                    v-if="filter"
                    @click="filter = ''"
                    class="table-component__filter__clear"
            >Ã—</a>
        </div>

        <div class="table-component__table-wrapper">
            <table :class="fullTableClass">
                <caption v-if="showCaption" class="table-component__table__caption" role="alert" aria-live="polite">
                    {{ ariaCaption }}
                </caption>
                <thead :class="fullTableHeadClass">
                <tr>
                    <table-column-header
                            @click="changeSorting"
                            v-for="column in columns"
                            :key="column.show"
                            :sort="sort"
                            :column="column"
                    ></table-column-header>
                </tr>
                </thead>
                <tbody :class="fullTableBodyClass">
                <table-row
                        v-for="row in displayedRows"
                        :key="row.vueTableComponentInternalRowId"
                        :row="row"
                        :columns="columns"
						:class="fullTableRowClass(row)"
						@rowClick="emitRowClick"
                ></table-row>
                </tbody>
                <tfoot>
                    <slot name="tfoot" :rows="rows"></slot>
                </tfoot>
            </table>
        </div>

        <div v-if="displayedRows.length === 0" class="table-component__message">
            {{ filterNoResults }}
        </div>


        <div style="display:none;">
            <slot></slot>
        </div>

        <pagination v-if="pagination" :pagination="pagination" @pageChange="pageChange"></pagination>
    </div>
</template>

<script>
    import Column from '../classes/Column';
    import expiringStorage from '../expiring-storage';
    import Row from '../classes/Row';
    import TableColumnHeader from './TableColumnHeader';
    import TableRow from './TableRow';
    import settings from '../settings';
    import Pagination from './Pagination';
    import { classList, pick } from '../helpers';

    export default {
        components: {
            TableColumnHeader,
            TableRow,
            Pagination,
        },

        props: {
            data: { default: () => [], type: [Array, Function] },

            showFilter: { default: true },
            showCaption: { default: true },

            sortBy: { default: '', type: String },
            sortOrder: { default: '', type: String },

            cacheKey: { default: null },
            cacheLifetime: { default: 5 },

            tableClass: { default: settings.tableClass },
            theadClass: { default: settings.theadClass },
			tbodyClass: { default: settings.tbodyClass },
			rowClass:  { default: settings.rowClass },
            filterInputClass: { default: settings.filterInputClass },
            filterPlaceholder: { default: settings.filterPlaceholder },
            filterNoResults: { default: settings.filterNoResults },
        },

        data: () => ({
            columns: [],
            rows: [],
            filter: '',
            sort: {
                fieldName: '',
                order: '',
            },
            pagination: null,

            localSettings: {},
        }),

        created() {
            this.sort.fieldName = this.sortBy;
            this.sort.order = this.sortOrder;

            this.restoreState();
        },

        async mounted() {
            const columnComponents = this.$slots.default
                .filter(column => column.componentInstance)
                .map(column => column.componentInstance);

            this.columns = columnComponents.map(
                column => new Column(column)
            );

            columnComponents.forEach(columnCom => {
                Object.keys(columnCom.$options.props).forEach(
                    prop => columnCom.$watch(prop, () => {
                        this.columns = columnComponents.map(
                            column => new Column(column)
                        );
                    })
                );
            });

            await this.mapDataToRows();
        },

        watch: {
            filter() {
                if (!this.usesLocalData) {
                    this.mapDataToRows();
                }

                this.saveState();
            },

            data() {
                if (this.usesLocalData) {
                    this.mapDataToRows();
                }
            },
        },

        computed: {
            fullTableClass() {
                return classList('table-component__table', this.tableClass);
            },

            fullTableHeadClass() {
                return classList('table-component__table__head', this.theadClass);
            },

            fullTableBodyClass() {
                return classList('table-component__table__body', this.tbodyClass);
            },

            fullFilterInputClass() {
                return classList('table-component__filter__field', this.filterInputClass);
            },

            ariaCaption() {
                if (this.sort.fieldName === '') {
                    return 'Table not sorted';
                }

                return `Table sorted by ${this.sort.fieldName} ` +
                    (this.sort.order === 'asc' ? '(ascending)' : '(descending)');
            },

            usesLocalData() {
                return Array.isArray(this.data);
            },

            displayedRows() {
                if (!this.usesLocalData) {
                    return this.sortedRows;
                }

                if (!this.showFilter) {
                    return this.sortedRows;
                }

                if (!this.columns.filter(column => column.isFilterable()).length) {
                    return this.sortedRows;
                }

                return this.sortedRows.filter(row => row.passesFilter(this.filter));
            },

            sortedRows() {
                if (!this.usesLocalData) {
                    return this.rows;
                }

                if (this.sort.fieldName === '') {
                    return this.rows;
                }

                if (this.columns.length === 0) {
                    return this.rows;
                }

                const sortColumn = this.getColumn(this.sort.fieldName);

                if (!sortColumn) {
                    return this.rows;
                }

                return this.rows.sort(sortColumn.getSortPredicate(this.sort.order, this.columns));
            },

            filterableColumnExists() {
                return this.columns.filter(c => c.isFilterable()).length > 0;
            },

            storageKey() {
                return this.cacheKey
                    ? `vue-table-component.${this.cacheKey}`
                    : `vue-table-component.${window.location.host}${window.location.pathname}${this.cacheKey}`;
            },
        },

        methods: {
            async pageChange(page) {
                this.pagination.currentPage = page;

                await this.mapDataToRows();
            },

            async mapDataToRows() {
                const data = this.usesLocalData
                    ? this.prepareLocalData()
                    : await this.fetchServerData();

                let rowId = 0;

                this.rows = data
                    .map(rowData => {
                        rowData.vueTableComponentInternalRowId = rowId++;
                        return rowData;
                    })
                    .map(rowData => new Row(rowData, this.columns));
            },

            prepareLocalData() {
                this.pagination = null;

                return this.data;
            },

            async fetchServerData() {
                const page = this.pagination && this.pagination.currentPage || 1;

                const response = await this.data({
                    filter: this.filter,
                    sort: this.sort,
                    page: page,
                });

                this.pagination = response.pagination;

                return response.data;
            },

            async refresh() {
                await this.mapDataToRows();
            },

            changeSorting(column) {
                if (this.sort.fieldName !== column.show) {
                    this.sort.fieldName = column.show;
                    this.sort.order = 'asc';
                } else {
                    this.sort.order = (this.sort.order === 'asc' ? 'desc' : 'asc');
                }

                if (!this.usesLocalData) {
                    this.mapDataToRows();
                }

                this.saveState();
            },

            getColumn(columnName) {
                return this.columns.find(column => column.show === columnName);
            },

            saveState() {
                expiringStorage.set(this.storageKey, pick(this.$data, ['filter', 'sort']), this.cacheLifetime);
            },

            restoreState() {
                const previousState = expiringStorage.get(this.storageKey);

                if (previousState === null) {
                    return;
                }

                this.sort = previousState.sort;
                this.filter = previousState.filter;

                this.saveState();
            },

			emitRowClick(row) {
				this.$emit('rowClick', row);
			},

            fullTableRowClass(row) {
				const classResult = typeof this.rowClass === 'function' ? `${this.rowClass(row)}` : this.rowClass
                return classList('table-component__table__body__row', classResult);
            },
		},
    };
</script>
