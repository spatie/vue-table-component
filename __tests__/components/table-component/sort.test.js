import TableComponent from '../../../src';
import Vue from 'vue/dist/vue.js';
import simulant from 'simulant';
import LocalStorageMock from '../../helpers/LocalStorageMock';

const localStorage = new LocalStorageMock();

window.localStorage = localStorage;

describe('Sortable tableComponent', () => {
    Vue.use(TableComponent);

    beforeEach(() => {
        localStorage.clear();
    });

    it('can sort the data with by a specific column', async () => {
        setDocumentInnerHtml({ sortBy: 'firstName' });

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('can sort the data with by another specific column', async () => {
        setDocumentInnerHtml({ sortBy: 'lastName' });

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('can sort the data with by a specific column in a specific order', async () => {
        setDocumentInnerHtml({ sortBy: 'firstName', order: 'desc' });

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('it will change the sort order when clicking the header of the column with the active sort', async () => {
        setDocumentInnerHtml({ sortBy: 'firstName' });

        await createVm();

        const firstColumnHeader = document.getElementsByTagName('th')[0];

        await simulant.fire(firstColumnHeader, 'click');

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('will sort the data ascendingly if the header of of column without the active sort is clicked', async () => {
        setDocumentInnerHtml({ sortBy: 'firstName' });

        await createVm();

        const secondColumnHeader = document.getElementsByTagName('th')[1];

        await simulant.fire(secondColumnHeader, 'click');

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('will not sort on a column that is not sortable', async () => {
        setDocumentInnerHtml({ sortBy: 'songs', order: 'desc' });

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('will not sort data when clicking a non-sortable column header', async () => {
        setDocumentInnerHtml({ sortBy: 'firstName' });

        await createVm();

        const thirdColumnHeader = document.getElementsByTagName('th')[2];

        await simulant.fire(thirdColumnHeader, 'click');

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('can sort on another column', async () => {
        document.body.innerHTML = `
            <div id="app">
                <div>
                    <table-component
                        :data="[{ firstName: 'John', songs: 30 },
                                { firstName: 'Paul', songs: 20 },
                                { firstName: 'George', songs: 420 },
                                { firstName: 'Ringo', songs: 210 }]"
                    >
                        <table-column show="firstName" label="First name" sort-by="songs"></table-column>
                        <table-column show="songs" data-type="numeric" label="Songs" sort-by="songs"></table-column>
                    </table-component>
                </div>
            </div>
        `;

        await createVm();

        const firstColumnHeader = document.getElementsByTagName('th')[0];

        await simulant.fire(firstColumnHeader, 'click');

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('will not sort rows in the footer', async () => {
        document.body.innerHTML = `
            <div id="app">
                <div>
                    <table-component
                        :data="[{ firstName: 'John', songs: 30 },
                                { firstName: 'Paul', songs: 20 },
                                { firstName: 'George', songs: 420 },
                                { firstName: 'Ringo', songs: 210 },
                                { firstName: 'Total Songs', songs: 680, isFooterRow: true }]"
                    >
                        <table-column show="firstName" label="First name" sort-by="songs"></table-column>
                        <table-column show="songs" data-type="numeric" label="Songs" sort-by="songs"></table-column>
                    </table-component>
                </div>
            </div>
        `;

        const table = await createVm();

        const firstColumnHeader = table.$el.getElementsByTagName('th')[0];

        await simulant.fire(firstColumnHeader, 'click');

        const body = table.$el.getElementsByTagName('tbody')[0];
        const footer = table.$el.getElementsByTagName('tfoot')[0];

        const bodyRows = body.getElementsByTagName('tr');
        const searchText = 'Total Songs';

        for (let i = 0; i < bodyRows.length; i++) {
            expect(findCellByText(bodyRows[i], searchText)).toBeFalsy();
        }

        expect(findCellByText(footer.children[0], searchText)).toBeTruthy();

        expect(document.body.innerHTML).toMatchSnapshot();
    });
});

function setDocumentInnerHtml({ sortBy = '', order = '' } = {}) {
    document.body.innerHTML = `
            <div id="app">
                <div>
                    <table-component
                        :data="[{ firstName: 'John', lastName: 'Lennon', songs: 72 },
                                { firstName: 'Paul', lastName: 'McCartney', songs: 70 },
                                { firstName: 'George', lastName: 'Harrison', songs: 22 },
                                { firstName: 'Ringo', lastName: 'Starr', songs: 2 }]"
                        sort-by="${sortBy}"
                        sort-order="${order}"
                    >
                        <table-column show="firstName" label="First name"></table-column>
                        <table-column show="lastName" label="Last name"></table-column>
                        <table-column show="songs" label="Songs" :sortable="false"></table-column>
                    </table-component>
                </div>
            </div>
        `;
}

function findCellByText(tableRow, text) {
    const tableRowArray = Array.from(tableRow.children);

    return tableRowArray.find(cell => {
        return cell.textContent === text;
    });
}

async function createVm() {
    const vm = new Vue({
        el: '#app',
    });

    await Vue.nextTick(() => {});

    const table = vm.$children[0];

    return table;
}
