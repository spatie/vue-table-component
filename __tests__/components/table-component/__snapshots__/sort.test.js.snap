// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sortable tableComponent can sort on another column 1`] = `

<div id="app">
  <div>
    <div class="table-component">
      <div class="table-component__filter">
        <input type="text"
               placeholder="Filter table…"
               class="table-component__filter__field"
        >
        <!---->
      </div>
      <div class="table-component__table-wrapper">
        <table class="table-component__table">
          <caption role="alert"
                   aria-live="polite"
                   class="table-component__table__caption"
          >
            Table sorted by firstName (ascending)
          </caption>
          <thead class="table-component__table__head">
            <tr>
              <th role="columnheader"
                  scope="col"
                  aria-sort="ascending"
                  class="table-component__th table-component__th--sort-asc"
              >
                First name
              </th>
              <th role="columnheader"
                  scope="col"
                  aria-sort="none"
                  class="table-component__th table-component__th--sort"
              >
                Songs
              </th>
            </tr>
          </thead>
          <tbody class="table-component__table__body">
            <tr>
              <td>
                Paul
              </td>
              <td>
                20
              </td>
            </tr>
            <tr>
              <td>
                John
              </td>
              <td>
                30
              </td>
            </tr>
            <tr>
              <td>
                Ringo
              </td>
              <td>
                210
              </td>
            </tr>
            <tr>
              <td>
                George
              </td>
              <td>
                420
              </td>
            </tr>
          </tbody>
          <tfoot>
          </tfoot>
        </table>
      </div>
      <!---->
      <div style="display: none;">
        <!---->
        <!---->
      </div>
      <!---->
    </div>
  </div>
</div>

`;

exports[`Sortable tableComponent can sort the data with by a specific column 1`] = `

<div id="app">
  <div>
    <div class="table-component">
      <div class="table-component__filter">
        <input type="text"
               placeholder="Filter table…"
               class="table-component__filter__field"
        >
        <!---->
      </div>
      <div class="table-component__table-wrapper">
        <table class="table-component__table">
          <caption role="alert"
                   aria-live="polite"
                   class="table-component__table__caption"
          >
            Table sorted by firstName (descending)
          </caption>
          <thead class="table-component__table__head">
            <tr>
              <th role="columnheader"
                  scope="col"
                  aria-sort="descending"
                  class="table-component__th table-component__th--sort-"
              >
                First name
              </th>
              <th role="columnheader"
                  scope="col"
                  aria-sort="none"
                  class="table-component__th table-component__th--sort"
              >
                Last name
              </th>
              <th role="columnheader"
                  scope="col"
                  aria-disabled="true"
                  class="table-component__th"
              >
                Songs
              </th>
            </tr>
          </thead>
          <tbody class="table-component__table__body">
            <tr>
              <td>
                George
              </td>
              <td>
                Harrison
              </td>
              <td>
                22
              </td>
            </tr>
            <tr>
              <td>
                John
              </td>
              <td>
                Lennon
              </td>
              <td>
                72
              </td>
            </tr>
            <tr>
              <td>
                Paul
              </td>
              <td>
                McCartney
              </td>
              <td>
                70
              </td>
            </tr>
            <tr>
              <td>
                Ringo
              </td>
              <td>
                Starr
              </td>
              <td>
                2
              </td>
            </tr>
          </tbody>
          <tfoot>
          </tfoot>
        </table>
      </div>
      <!---->
      <div style="display: none;">
        <!---->
        <!---->
        <!---->
      </div>
      <!---->
    </div>
  </div>
</div>

`;

exports[`Sortable tableComponent can sort the data with by a specific column in a specific order 1`] = `

<div id="app">
  <div>
    <div class="table-component">
      <div class="table-component__filter">
        <input type="text"
               placeholder="Filter table…"
               class="table-component__filter__field"
        >
        <!---->
      </div>
      <div class="table-component__table-wrapper">
        <table class="table-component__table">
          <caption role="alert"
                   aria-live="polite"
                   class="table-component__table__caption"
          >
            Table sorted by firstName (descending)
          </caption>
          <thead class="table-component__table__head">
            <tr>
              <th role="columnheader"
                  scope="col"
                  aria-sort="descending"
                  class="table-component__th table-component__th--sort-desc"
              >
                First name
              </th>
              <th role="columnheader"
                  scope="col"
                  aria-sort="none"
                  class="table-component__th table-component__th--sort"
              >
                Last name
              </th>
              <th role="columnheader"
                  scope="col"
                  aria-disabled="true"
                  class="table-component__th"
              >
                Songs
              </th>
            </tr>
          </thead>
          <tbody class="table-component__table__body">
            <tr>
              <td>
                Ringo
              </td>
              <td>
                Starr
              </td>
              <td>
                2
              </td>
            </tr>
            <tr>
              <td>
                Paul
              </td>
              <td>
                McCartney
              </td>
              <td>
                70
              </td>
            </tr>
            <tr>
              <td>
                John
              </td>
              <td>
                Lennon
              </td>
              <td>
                72
              </td>
            </tr>
            <tr>
              <td>
                George
              </td>
              <td>
                Harrison
              </td>
              <td>
                22
              </td>
            </tr>
          </tbody>
          <tfoot>
          </tfoot>
        </table>
      </div>
      <!---->
      <div style="display: none;">
        <!---->
        <!---->
        <!---->
      </div>
      <!---->
    </div>
  </div>
</div>

`;

exports[`Sortable tableComponent can sort the data with by another specific column 1`] = `

<div id="app">
  <div>
    <div class="table-component">
      <div class="table-component__filter">
        <input type="text"
               placeholder="Filter table…"
               class="table-component__filter__field"
        >
        <!---->
      </div>
      <div class="table-component__table-wrapper">
        <table class="table-component__table">
          <caption role="alert"
                   aria-live="polite"
                   class="table-component__table__caption"
          >
            Table sorted by lastName (descending)
          </caption>
          <thead class="table-component__table__head">
            <tr>
              <th role="columnheader"
                  scope="col"
                  aria-sort="none"
                  class="table-component__th table-component__th--sort"
              >
                First name
              </th>
              <th role="columnheader"
                  scope="col"
                  aria-sort="descending"
                  class="table-component__th table-component__th--sort-"
              >
                Last name
              </th>
              <th role="columnheader"
                  scope="col"
                  aria-disabled="true"
                  class="table-component__th"
              >
                Songs
              </th>
            </tr>
          </thead>
          <tbody class="table-component__table__body">
            <tr>
              <td>
                George
              </td>
              <td>
                Harrison
              </td>
              <td>
                22
              </td>
            </tr>
            <tr>
              <td>
                John
              </td>
              <td>
                Lennon
              </td>
              <td>
                72
              </td>
            </tr>
            <tr>
              <td>
                Paul
              </td>
              <td>
                McCartney
              </td>
              <td>
                70
              </td>
            </tr>
            <tr>
              <td>
                Ringo
              </td>
              <td>
                Starr
              </td>
              <td>
                2
              </td>
            </tr>
          </tbody>
          <tfoot>
          </tfoot>
        </table>
      </div>
      <!---->
      <div style="display: none;">
        <!---->
        <!---->
        <!---->
      </div>
      <!---->
    </div>
  </div>
</div>

`;

exports[`Sortable tableComponent it will change the sort order when clicking the header of the column with the active sort 1`] = `

<div id="app">
  <div>
    <div class="table-component">
      <div class="table-component__filter">
        <input type="text"
               placeholder="Filter table…"
               class="table-component__filter__field"
        >
        <!---->
      </div>
      <div class="table-component__table-wrapper">
        <table class="table-component__table">
          <caption role="alert"
                   aria-live="polite"
                   class="table-component__table__caption"
          >
            Table sorted by firstName (ascending)
          </caption>
          <thead class="table-component__table__head">
            <tr>
              <th role="columnheader"
                  scope="col"
                  aria-sort="ascending"
                  class="table-component__th table-component__th--sort-asc"
              >
                First name
              </th>
              <th role="columnheader"
                  scope="col"
                  aria-sort="none"
                  class="table-component__th table-component__th--sort"
              >
                Last name
              </th>
              <th role="columnheader"
                  scope="col"
                  aria-disabled="true"
                  class="table-component__th"
              >
                Songs
              </th>
            </tr>
          </thead>
          <tbody class="table-component__table__body">
            <tr>
              <td>
                George
              </td>
              <td>
                Harrison
              </td>
              <td>
                22
              </td>
            </tr>
            <tr>
              <td>
                John
              </td>
              <td>
                Lennon
              </td>
              <td>
                72
              </td>
            </tr>
            <tr>
              <td>
                Paul
              </td>
              <td>
                McCartney
              </td>
              <td>
                70
              </td>
            </tr>
            <tr>
              <td>
                Ringo
              </td>
              <td>
                Starr
              </td>
              <td>
                2
              </td>
            </tr>
          </tbody>
          <tfoot>
          </tfoot>
        </table>
      </div>
      <!---->
      <div style="display: none;">
        <!---->
        <!---->
        <!---->
      </div>
      <!---->
    </div>
  </div>
</div>

`;

exports[`Sortable tableComponent will not sort data when clicking a non-sortable column header 1`] = `

<div id="app">
  <div>
    <div class="table-component">
      <div class="table-component__filter">
        <input type="text"
               placeholder="Filter table…"
               class="table-component__filter__field"
        >
        <!---->
      </div>
      <div class="table-component__table-wrapper">
        <table class="table-component__table">
          <caption role="alert"
                   aria-live="polite"
                   class="table-component__table__caption"
          >
            Table sorted by firstName (descending)
          </caption>
          <thead class="table-component__table__head">
            <tr>
              <th role="columnheader"
                  scope="col"
                  aria-sort="descending"
                  class="table-component__th table-component__th--sort-"
              >
                First name
              </th>
              <th role="columnheader"
                  scope="col"
                  aria-sort="none"
                  class="table-component__th table-component__th--sort"
              >
                Last name
              </th>
              <th role="columnheader"
                  scope="col"
                  aria-disabled="true"
                  class="table-component__th"
              >
                Songs
              </th>
            </tr>
          </thead>
          <tbody class="table-component__table__body">
            <tr>
              <td>
                George
              </td>
              <td>
                Harrison
              </td>
              <td>
                22
              </td>
            </tr>
            <tr>
              <td>
                John
              </td>
              <td>
                Lennon
              </td>
              <td>
                72
              </td>
            </tr>
            <tr>
              <td>
                Paul
              </td>
              <td>
                McCartney
              </td>
              <td>
                70
              </td>
            </tr>
            <tr>
              <td>
                Ringo
              </td>
              <td>
                Starr
              </td>
              <td>
                2
              </td>
            </tr>
          </tbody>
          <tfoot>
          </tfoot>
        </table>
      </div>
      <!---->
      <div style="display: none;">
        <!---->
        <!---->
        <!---->
      </div>
      <!---->
    </div>
  </div>
</div>

`;

exports[`Sortable tableComponent will not sort on a column that is not sortable 1`] = `

<div id="app">
  <div>
    <div class="table-component">
      <div class="table-component__filter">
        <input type="text"
               placeholder="Filter table…"
               class="table-component__filter__field"
        >
        <!---->
      </div>
      <div class="table-component__table-wrapper">
        <table class="table-component__table">
          <caption role="alert"
                   aria-live="polite"
                   class="table-component__table__caption"
          >
            Table sorted by songs (descending)
          </caption>
          <thead class="table-component__table__head">
            <tr>
              <th role="columnheader"
                  scope="col"
                  aria-sort="none"
                  class="table-component__th table-component__th--sort"
              >
                First name
              </th>
              <th role="columnheader"
                  scope="col"
                  aria-sort="none"
                  class="table-component__th table-component__th--sort"
              >
                Last name
              </th>
              <th role="columnheader"
                  scope="col"
                  aria-disabled="true"
                  class="table-component__th"
              >
                Songs
              </th>
            </tr>
          </thead>
          <tbody class="table-component__table__body">
            <tr>
              <td>
                John
              </td>
              <td>
                Lennon
              </td>
              <td>
                72
              </td>
            </tr>
            <tr>
              <td>
                Paul
              </td>
              <td>
                McCartney
              </td>
              <td>
                70
              </td>
            </tr>
            <tr>
              <td>
                George
              </td>
              <td>
                Harrison
              </td>
              <td>
                22
              </td>
            </tr>
            <tr>
              <td>
                Ringo
              </td>
              <td>
                Starr
              </td>
              <td>
                2
              </td>
            </tr>
          </tbody>
          <tfoot>
          </tfoot>
        </table>
      </div>
      <!---->
      <div style="display: none;">
        <!---->
        <!---->
        <!---->
      </div>
      <!---->
    </div>
  </div>
</div>

`;

exports[`Sortable tableComponent will sort the data ascendingly if the header of of column without the active sort is clicked 1`] = `

<div id="app">
  <div>
    <div class="table-component">
      <div class="table-component__filter">
        <input type="text"
               placeholder="Filter table…"
               class="table-component__filter__field"
        >
        <!---->
      </div>
      <div class="table-component__table-wrapper">
        <table class="table-component__table">
          <caption role="alert"
                   aria-live="polite"
                   class="table-component__table__caption"
          >
            Table sorted by lastName (ascending)
          </caption>
          <thead class="table-component__table__head">
            <tr>
              <th role="columnheader"
                  scope="col"
                  aria-sort="none"
                  class="table-component__th table-component__th--sort"
              >
                First name
              </th>
              <th role="columnheader"
                  scope="col"
                  aria-sort="ascending"
                  class="table-component__th table-component__th--sort-asc"
              >
                Last name
              </th>
              <th role="columnheader"
                  scope="col"
                  aria-disabled="true"
                  class="table-component__th"
              >
                Songs
              </th>
            </tr>
          </thead>
          <tbody class="table-component__table__body">
            <tr>
              <td>
                George
              </td>
              <td>
                Harrison
              </td>
              <td>
                22
              </td>
            </tr>
            <tr>
              <td>
                John
              </td>
              <td>
                Lennon
              </td>
              <td>
                72
              </td>
            </tr>
            <tr>
              <td>
                Paul
              </td>
              <td>
                McCartney
              </td>
              <td>
                70
              </td>
            </tr>
            <tr>
              <td>
                Ringo
              </td>
              <td>
                Starr
              </td>
              <td>
                2
              </td>
            </tr>
          </tbody>
          <tfoot>
          </tfoot>
        </table>
      </div>
      <!---->
      <div style="display: none;">
        <!---->
        <!---->
        <!---->
      </div>
      <!---->
    </div>
  </div>
</div>

`;

exports[`Sortable tableComponent wont break if a sortable column has no data 1`] = `

<div id="app">
  <div>
    <div class="table-component">
      <div class="table-component__filter">
        <input type="text"
               placeholder="Filter table…"
               class="table-component__filter__field"
        >
        <!---->
      </div>
      <div class="table-component__table-wrapper">
        <table class="table-component__table">
          <caption role="alert"
                   aria-live="polite"
                   class="table-component__table__caption"
          >
            Table sorted by lastName (ascending)
          </caption>
          <thead class="table-component__table__head">
            <tr>
              <th role="columnheader"
                  scope="col"
                  aria-sort="none"
                  class="table-component__th table-component__th--sort"
              >
                First name
              </th>
              <th role="columnheader"
                  scope="col"
                  aria-sort="ascending"
                  class="table-component__th table-component__th--sort-asc"
              >
                Last name
              </th>
              <th role="columnheader"
                  scope="col"
                  aria-sort="none"
                  class="table-component__th table-component__th--sort"
              >
                Songs
              </th>
            </tr>
          </thead>
          <tbody class="table-component__table__body">
            <tr>
              <td>
                John
              </td>
              <td>
              </td>
              <td>
                30
              </td>
            </tr>
            <tr>
              <td>
                Paul
              </td>
              <td>
              </td>
              <td>
                20
              </td>
            </tr>
            <tr>
              <td>
                George
              </td>
              <td>
              </td>
              <td>
                420
              </td>
            </tr>
            <tr>
              <td>
                Ringo
              </td>
              <td>
              </td>
              <td>
                210
              </td>
            </tr>
          </tbody>
          <tfoot>
          </tfoot>
        </table>
      </div>
      <!---->
      <div style="display: none;">
        <!---->
        <!---->
        <!---->
      </div>
      <!---->
    </div>
  </div>
</div>

`;
